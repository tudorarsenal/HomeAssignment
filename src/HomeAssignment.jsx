import { useState, useEffect } from 'react';

import Table from './components/Table/Table';
import { getDevices } from './utils/apis';
import { ERROR_MESSAGE_DOWNLOAD, NO_DEVICES_DOWNLOAD } from './utils/textsEn';

import './HomeAssignment.scss';

const tableHeaders = ['Name', 'Device', 'Path', 'Status'];

const downloadRules = {
  disableButton: false,
  rule: (device) => device.status === 'available',
  errorMessage: ERROR_MESSAGE_DOWNLOAD,
};

const HomeAssignment = () => {
  const [devices, setDevices] = useState([]);

  useEffect(() => {
    getDeviceData();
  }, []);

  const getDeviceData = () => {
    try {
      const devicesData = getDevices();

      setDevices(devicesData);
    } catch (error) {
      console.log('error', error);
    }
  };

  // display device and path of the selected devices matching the rule
  const handleDownload = (selectedDevices) => {
    let finalMessage = '';

    selectedDevices?.forEach((row) => {
      const currentDevice = devices?.find((device) => device?.name === row);

      if (downloadRules?.rule(currentDevice)) {
        finalMessage += `${
          currentDevice?.device + ': ' + currentDevice?.path
        } \n`;
      }
    });

    if (!finalMessage) {
      finalMessage = NO_DEVICES_DOWNLOAD;
    }

    alert(finalMessage);
  };

  return (
    <div className="home-assignment">
      <Table
        headers={tableHeaders}
        data={devices}
        isDownloadable={true}
        handleDownload={handleDownload}
        downloadRules={downloadRules}
      />
    </div>
  );
};

export default HomeAssignment;
