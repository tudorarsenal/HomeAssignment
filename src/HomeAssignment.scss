.home-assignment {
  margin: 100px;
}
