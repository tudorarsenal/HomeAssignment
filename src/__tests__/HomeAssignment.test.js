import renderer from 'react-test-renderer';

import HomeAssignment from '../HomeAssignment';

describe('HomeAssignment', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<HomeAssignment />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
