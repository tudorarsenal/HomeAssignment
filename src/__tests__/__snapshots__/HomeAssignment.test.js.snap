// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeAssignment renders correctly 1`] = `
<div
  className="home-assignment"
>
  <div
    className="table"
  >
    <div
      className="table-actions"
    >
      <div
        className="table-actions__select"
      >
        <label
          className="visually-hidden"
          htmlFor="select-all"
        >
          Select all
        </label>
        <input
          checked={true}
          className="table-actions__check-all-indeterminate"
          data-testid="select-all-checkbox"
          id="select-all"
          name="select-all"
          onChange={[Function]}
          type="checkbox"
        />
        <div
          className="table-actions__selection-count"
        >
          None Selected
        </div>
      </div>
    </div>
    <div
      className="table-header table-header-selectable"
    >
      <div
        className="table-header__item table-header__name"
      >
        Name
      </div>
      <div
        className="table-header__item table-header__device"
      >
        Device
      </div>
      <div
        className="table-header__item table-header__path"
      >
        Path
      </div>
      <div
        className="table-header__item table-header__status"
      >
        Status
      </div>
    </div>
    <div
      className="table-body__no-data"
    >
      No data
    </div>
  </div>
</div>
`;
